src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,398 @@
+        for tgt, deps in list(self._raw_deps.items()):
+            for src in deps:
+                if src not in self.node_meta:
+                    self.node_meta[src] = {
+                        "kind": "state",
+                        "agent": self.block_agent or "global",
+                        "plate": self.block_agent,
+                        "observed": False,
+                    }
+
+        self._identify_time_dependencies()
+        self._assemble_edges()
+        self._collect_formulas()
+        self._collect_plates()
+        self._add_lag_variable_metadata()
+        return self
+
+    def _collect_nodes(self):
+        """Classify every variable and record its metadata."""
+
+        for blk in self._blocks:
+            # Use block_agent if specified, otherwise check if block has agent attribute
+            block_plate = self.block_agent or getattr(blk, "agent", None)
+
+            # 1) Collect shocks - typically global/exogenous, not in any plate
+            for var, shock_def in blk.shocks.items():
+                self.node_meta[var] = {
+                    "kind": "shock",
+                    "agent": "global",
+                    "plate": None,  # Shocks are not in any plate
+                    "observed": False,
+                }
+
+            # 2) Collect dynamics - all belong to the block's plate
+            for var, rule in blk.dynamics.items():
+                if isinstance(rule, Control):
+                    kind = "control"
+                    # Control might specify its own agent for other purposes
+                    agent = getattr(rule, "agent", None) or block_plate or "global"
+                else:
+                    kind = "state"
+                    agent = block_plate or "global"
+
+                self.node_meta[var] = {
+                    "kind": kind,
+                    "agent": agent,
+                    "plate": block_plate,  # All dynamics belong to the block's plate
+                    "observed": (kind == "control") or (var in self.observables),
+                }
+
+            # 3) Collect rewards - belong to the block's plate
+            for var, rule in blk.reward.items():
+                self.node_meta[var] = {
+                    "kind": "reward",
+                    "agent": block_plate or "global",
+                    "plate": block_plate,  # Rewards belong to the block's plate
+                    "observed": True,  # Rewards are always observed
+                }
+
+        # 4) Collect parameters - global, not in any plate
+        defined_vars = set(self.node_meta.keys())
+        for p in self.calibration:
+            if p not in defined_vars:
+                self.node_meta[p] = {
+                    "kind": "param",
+                    "agent": "global",
+                    "plate": None,  # Parameters are not in any plate
+                    "observed": False,
+                }
+
+    def _extract_dependencies(self, rule, var_name=None):
```

Comment:
var_name is never used. It says it's optional and 'for error messages' in the docs, but that is misleading, since no error message is reported.

I would remove this argument.